### ISSUE TITLE
Aggregate queries

### ISSUE THREAD

**Issue by tr11** [bokwoon95/sq#6]
I've been trying to create a dynamic query to do a simple count as:
```
sq.Postgres.From(tbl).Select(sq.CountStar().As("count"))
```
The SQL generated from a `CompileExec` call is correct:
```
SELECT COUNT(*) AS count FROM table
```
but I can't seem to retrieve the result with a `rowmapper` function. I'm using for `rowmapper` 
```
func(row *sq.Row) int {
	return row.Int("count")
}
```
but the fetch call fails with an error (`column "count" does not exist (SQLSTATE 42703)`).

What's the correct way to do aggregations at the top level without specifying columns and group by clauses?

Thanks!

**Comment by bokwoon95** [comment]
Oh the Fetch function doesn't work like that, it takes whatever you feed into row.Int() and inserts it into the SELECT clause of the query. So you thought you were running `SELECT COUNT(*) AS count FROM table` but actually it's running `SELECT count FROM table` (which is why it's complaining that the column "count" does not exist). To see the generated query being run by Fetch(), you can [wrap the DB passed in with sq.Log()](https://bokwoon.neocities.org/sq#logging) e.g.

```go
tbl := sq.New[ACTOR]("a")
// With logging                    ↓ wrap the db
firstName, err := sq.FetchOne(sq.Log(db),
    sq.Postgres.From(tbl).Where(tbl.LAST_NAME.In([]string{"AKROYD", "ALLEN", "WILLIAMS"})),
    func(row *sq.Row) string {
        return row.String("first_name")
    },
)
```
```
2022/02/06 15:34:36 [OK] SELECT first_name FROM actor AS a WHERE a.last_name IN ('AKROYD', 'ALLEN', 'WILLIAMS') | timeTaken=9.834µs rowCount=9
```

To answer your question, you shouldn't use Select() when calling Fetch, because whatever needs to be SELECT-ed is automatically obtained from the rowmapper:

```go
count, err := sq.FetchOne(sq.Log(db),
    sq.Postgres.From(tbl),
    func(row *sq.Row) int {
        return row.Int("COUNT(*) AS count")
    },
)
```

**Comment by esiqveland** [comment]
> To answer your question, you shouldn't use Select() when calling Fetch, because whatever needs to be SELECT-ed is obtained from the rowmapper and overwrites whatever fields you have previously passed to Select(). Just leave it out:
> 
> ```go
> count, err := sq.FetchOne(sq.Log(db),
>     sq.Postgres.From(tbl),
>     func(row *sq.Row) int {
>         return row.Int("COUNT(*) AS count")
>     },
> )
> ```

Sorry for reviving an old thread. Is there a way to use the RowMapper function without this query writing?

I really like the rowmapper function, I just would prefer to use it without modifying the select query. Is that possible with `sq`? It would really help to ease it into existing codebases.

**Comment by bokwoon95** [comment]
Can you show a code example of what you mean? What do you envision the API to look like?

**Comment by esiqveland** [comment]
Yes, sorry, not the best example.

Using the mapper function seems to imply you also must use the automatic query rewriting that expands the `SELECT {*}` part with some 'magic'. It also requires that users know about the fact that the mapper function can _not_ use branching while querying column names, and that the mapper function will be called with an empty row before the query.
I think it would be simpler to introduce `sq` if we could use the mapper function without these special features.

It would then require you to specify all columns in the SELECT directly. It would look like:

```go
list, err := sq.FetchAll(
   sq.Queryf("SELECT id, metadata, created_at FROM events WHERE id = {id}", sql.Named("id", id))
    sq.Log(db),
    func(row *sq.Row) MyEvent {
        // no query rewrite needed:
        return MyEvent{
            ID: parseId(row.String("id")),
            Metadata: parseMetadata(row.Bytes("metadata")),
            CreatedAt: row.Time("created_at"),
        }
    },
)

```

**Comment by bokwoon95** [comment]
Yeah, I think this feature makes sense. Let me implement it

**Comment by bokwoon95** [comment]
This has been implemented in v0.5.0, check the documentation here: [https://bokwoon.neocities.org/sq#static-vs-dynamic-queries](https://bokwoon.neocities.org/sq#static-vs-dynamic-queries).
